packages/typescript-estree/src/create-program/createOrUpdateProjectService.ts

```diff
@@ -0,0 +1,94 @@
+import * as tsserver from 'typescript/lib/tsserverlibrary';
+
+const doNothing = (): void => {};
+
+const createStubFileWatcher = (): tsserver.FileWatcher => ({
+  close: doNothing,
+});
+
+const SEEN_PROJECTS = new Set<string>([]);
+
+let TSSERVER_PROJECT_SERVICE: ReturnType<
+  typeof createOrUpdateProjectService
+> | null = null;
+
+export function clearTSServerProjectService(): void {
+  SEEN_PROJECTS.clear();
+  TSSERVER_PROJECT_SERVICE = null;
+}
+
+export function createOrUpdateProjectService(
+  codeFullText: string,
+  filePath: string,
+  projectConfigFiles: string[] | null,
+): tsserver.server.ProjectService {
+  TSSERVER_PROJECT_SERVICE ??= createProjectService();
+
+  if (projectConfigFiles) {
+    for (const projectFileName of projectConfigFiles) {
+      if (!SEEN_PROJECTS.has(projectFileName)) {
+        SEEN_PROJECTS.add(projectFileName);
+        TSSERVER_PROJECT_SERVICE.openExternalProject({
+          options: {},
+          projectFileName,
+          rootFiles: [
+            {
+              content: codeFullText,
+              fileName: filePath,
+            },
+          ],
+        });
+      }
+    }
+  }
```

Comment:
When `project` is set to something other than `./tsconfig.json`, the project service doesn't seem to natively understand to ask for it. You can see this in unit tests such as `no-unsafe-call.test.ts` (any `RuleTester` with a `parserOptions.project` like `./tsconfig.*.json`). [[Example CI failure](https://github.com/typescript-eslint/typescript-eslint/actions/runs/4800211437/jobs/8540946706)]

```plaintext
    Message:
      A fatal parsing error occurred: Parsing error: Cannot read file '/home/runner/work/typescript-eslint/typescript-eslint/packages/eslint-plugin/tests/fixtures/tsconfig.noImplicitThis.json'
```

It feels very weird to specifically indicate that a single file name & content should match for the project. But calling `openExternalProject` caused a significant slowdown in unit tests in `eslint-plugin`: `time TYPESCRIPT_ESLINT_EXPERIMENTAL_TSSERVER=true yarn jest naming-convention` went from ~90s to ~120s on my Mac Studio.

Is there a better strategy for what we're trying to do?

Reply:
My impression is still that there's no reason to respect `project` set to any path when run with the project service, because it's not going to be respectd in VS Code either. Adding the project service means that no configuration is needed at all for any tsconfigs, besides one toggle to say whether TS is loaded or not.

Am I missing something?

Reply:
We do need to respect some compiler options in our lint rules though. Some lint rules behave differently when e.g. `noImplicitThis` or `strictNullChecks` are enabled. `/Users/josh/repos/typescript-eslint/packages/eslint-plugin/tests/rules/non-nullable-type-assertion-style.test.ts` shows tests that are specifically set up for `noUncheckedIndexedAccess`.

Separately, if someone is using a `tsconfig.eslint.json` to set up different `paths` or other significant changes, we need to respect that in our rules. One common strategy is to have a `tsconfig.eslint.json` that specifies `allowJs: true` to typed rules are allowed to run on `.eslintrc.js` & other config files.

Reply:
> We do need to respect some compiler options in our lint rules though. Some lint rules behave differently when e.g. `noImplicitThis` or `strictNullChecks` are enabled. `/Users/josh/repos/typescript-eslint/packages/eslint-plugin/tests/rules/non-nullable-type-assertion-style.test.ts` shows tests that are specifically set up for `noUncheckedIndexedAccess`.

Is this just exclusive for your testing?

> Separately, if someone is using a `tsconfig.eslint.json` to set up different `paths` or other significant changes, we need to respect that in our rules. One common strategy is to have a `tsconfig.eslint.json` that specifies `allowJs: true` to typed rules are allowed to run on `.eslintrc.js` & other config files.

The former is the thing that using the project service solves, though; you don't need to trick TS into loading the code into one project anymore.

Loading js, yeah, not sure there.

Reply:
> Is this just exclusive for your testing?

The testing is reflective of real users. We've seen cases where people set varying TSConfig options in different parts of their projects -sometimes in a formalized monorepo, sometimes in a large vaguely-organized single repo- and want that respected in their lint rules.

> add some sort of config for the "implicit" project used for files that aren't in any other project?

Interesting - like a default project that always exists, provided by our tooling? Based on `tsconfig.json`? But what if the user doesn't have a `tsconfig.json` visible (e.g. is linting from a subdirectory)?

Reply:
> The testing is reflective of real users. We've seen cases where people set varying TSConfig options in different parts of their projects -sometimes in a formalized monorepo, sometimes in a large vaguely-organized single repo- and want that respected in their lint rules.

Right, but if they have that, then the tsconfigs on disk should already reflect that setup, right? So they can typecheck and get tsserver working?

> Interesting - like a default project that always exists, provided by our tooling? Based on `tsconfig.json`? But what if the user doesn't have a `tsconfig.json` visible (e.g. is linting from a subdirectory)?

So the implicit project is what you're using when you open `.eslintrc.js` in a project with a tsconfig that doesn't include it. All code has to be in some project in order to analyze it. So, I would expect that with the ProjectService, you can open any file and get results.

Whether or not the defualt settings are good, i'm not sure. VS Code configures it differently: https://github.com/microsoft/vscode/blob/2d28a86e1d8cb556d06c1f646ca7370ed761e5b0/extensions/typescript-language-features/src/tsconfig.ts#L30

Reply:
We talked on Discord - I'll post a summary soon. tl;dr is - I'm going to prioritize getting this integration simple & shippable, perhaps to the point of excluding / ignoring these unit tests.

Reply:
Ok! Summarizing our chat:
* I (Josh) have been trying to come up with repro cases showing why users would want to have their own `project`(s) (`tsconfig.eslint.json`, `tsconfig.differentCompilerOption`, ...) - and have yet to come up with a broadly compelling reason 
    * `allowJs: true` is given to us for free by virtue of the default inferred TS project (see the vscode link 👆)
    * Even if we had a great use case, it would be an _unusual_ and _weird_ one: because it would mean people are running ESLint in ways different from how the editor views files!
* The vast majority of users who would want behavior akin to today's `project: true` would likely be happy with `useProjectService`. 
    * If we do want to use an API for adding manual non-inferred projects (e.g. `openExternalProject`) then we can always add it in later - it's not necessary for an MVP
* My (Josh) next steps will be to:
    1. Remove the `openExternalProject` usage
    2. Try to work around it in unit tests, and fall back to skipping tests that can't
    3. Un-draft this PR for review
    4. _(hopeful)_ get this PR release & available
    5. _(hopeful)_ update docs to mention this experimental project service